### PR TITLE
Dev improvements

### PR DESCRIPTION
**Pull request by jessicabrinegar**
Implemented recommended changes from previous pull request (seen in Aug 5th & 6th commits)
1) Only one type for account, and one for transaction
2) Validations in one pipe (instead of two) b/c I was able to access the route method (using the request object)
3) Validations within the pipe are broken into several functions, improving readability
4) When data is updated, a new object is created & returned (improved immutable data manipulation)
5) Withdraw & deposit service functions were combined into one function due to repeated code
6) The pipe transforms the request body instead of having the transformation within the controller (adding id & account_id fields so they don't have to be included in the request)

Thank you for your review.

### CODE REVIEW

**Comment by jessicabrinegar** [comment]
@ifielder 
I didn't write any logic for currency conversion or rules around sticking to a specified currency once account is created. Figured I'd save the time to learn more Nest specific things, but can totally add that in if you want me to.

**Comment by ifielder** [comment]
> @ifielder 
> 
> I didn't write any logic for currency conversion or rules around sticking to a specified currency once account is created. Figured I'd save the time to learn more Nest specific things, but can totally add that in if you want me to.

Yes, focus on the nest specifics and the whys are definitely the way to go 🔥

File: src/transactions/transactions.controller.spec.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import { Test, TestingModule } from '@nestjs/testing';
```
Comment (cyrusBaker) [inline src/transactions/transactions.controller.spec.ts:1]: I'm not sure if testing is part of this project or not, but we typically don't write unit tests for the controllers. We write them for services, entities, and mappers, and we also do end-to-end testing. But you will learn more about that as you get into domain driven design.

**Review by cyrusBaker — APPROVED** [review]

File: src/transactions/transactions.controller.spec.ts (hunk 1 above)
Comment (jessicabrinegar) [reply src/transactions/transactions.controller.spec.ts:1]: That's good to know, thank you! I actually didn't mess with any of the testing files at all.. they were just created when I bootstrapped the project. It is something that I need/want to learn, though.

File: src/accounts/accounts.service.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        }
+        return account;
+    }
+
+    create(account: AccountDto): AccountDto {
+        this.accounts.push(account);
+        return account;
+    }
+
+    updateAccount(id: string, data: Partial<AccountDto>): AccountDto {
```
Comment (ifielder) [inline src/accounts/accounts.service.ts:26]: For a little more readability you can reduce the use of the index and use the map and find methods 


```
updateAccount(id: string, data: Partial<AccountDto>): AccountDto {
    const accountToUpdate = this.accounts.find(account => account.id === id);
    if (!accountToUpdate) {
        throw new NotFoundException(`No account found with ID ${id}`);
    }
    const updatedAccount = { ...accountToUpdate, ...data };
    
    this.accounts = this.accounts.map(account => 
        account.id === id ? updatedAccount : account
    );
    
    return updatedAccount;
}


```

File: src/accounts/accounts.service.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        const updatedAccount = { ...this.accounts[index], ...data};
+        this.accounts = [
+            ...this.accounts.slice(0, index),
+            updatedAccount,
+            ...this.accounts.slice(index + 1)
+        ];
+        return updatedAccount;
+    }
+
+    updateBalance(id: string, amount: number, type: string): { newBalance: number } {
```
Comment (ifielder) [inline src/accounts/accounts.service.ts:41]: Ideally, type would be a typescript type so that we can only allow certain types send as an argument !

**Review by ifielder — APPROVED** [review]
Great work @jessicabrinegar

This is good to merge! I just left a couple small suggestions

**Pull request merged**